ethclient/ethclient.go

```diff
@@ -0,0 +1,586 @@
+	var json *rpcTransaction
+	err := ec.c.CallContext(ctx, &json, "eth_getTransactionByBlockHashAndIndex", blockHash, hexutil.Uint64(index))
+	if err != nil {
+		return nil, err
+	}
+	if json == nil {
+		return nil, coreth.NotFound
+	} else if _, r, _ := json.tx.RawSignatureValues(); r == nil {
+		return nil, fmt.Errorf("server returned transaction without signature")
+	}
+	if json.From != nil && json.BlockHash != nil {
+		setSenderFromServer(json.tx, *json.From, *json.BlockHash)
+	}
+	return json.tx, err
+}
+
+// TransactionReceipt returns the receipt of a transaction by transaction hash.
+// Note that the receipt is not available for pending transactions.
+func (ec *Client) TransactionReceipt(ctx context.Context, txHash common.Hash) (*types.Receipt, error) {
+	var r *types.Receipt
+	err := ec.c.CallContext(ctx, &r, "eth_getTransactionReceipt", txHash)
+	if err == nil {
+		if r == nil {
+			return nil, coreth.NotFound
+		}
+	}
+	return r, err
+}
+
+func toBlockNumArg(number *big.Int) string {
+	// The Ethereum implementation uses a different mapping from
+	// negative numbers to special strings (latest, pending) then is
+	// used on its server side. See rpc/types.go for the comparison.
+	// if number == nil {
+	// 	return "latest"
+	// }
+	// pending := big.NewInt(-1)
+	// if number.Cmp(pending) == 0 {
+	// 	return "pending"
+	// }
+
+	// In Coreth, latest, pending, and accepted are all treated the same
+	// therefore, if [number] is nil or a negative number in [-3, -1]
+	// we want the latest accepted block
+	if number == nil {
+		return "latest"
+	}
+	low := big.NewInt(-3)
+	high := big.NewInt(-1)
+	if number.Cmp(low) >= 0 && number.Cmp(high) <= 0 {
+		return "latest"
+	}
+	return hexutil.EncodeBig(number)
+}
+
+type rpcProgress struct {
+	StartingBlock hexutil.Uint64
+	CurrentBlock  hexutil.Uint64
+	HighestBlock  hexutil.Uint64
+	PulledStates  hexutil.Uint64
+	KnownStates   hexutil.Uint64
+}
+
+// SyncProgress retrieves the current progress of the sync algorithm. If there's
+// no sync currently running, it returns nil.
+// eth_syncing is not implemented in Coreth
+// func (ec *Client) SyncProgress(ctx context.Context) (*coreth.SyncProgress, error) {
+// 	var raw json.RawMessage
+// 	if err := ec.c.CallContext(ctx, &raw, "eth_syncing"); err != nil {
+// 		return nil, err
+// 	}
+// 	// Handle the possible response types
+// 	var syncing bool
+// 	if err := json.Unmarshal(raw, &syncing); err == nil {
+// 		return nil, nil // Not syncing (always false)
+// 	}
+// 	var progress *rpcProgress
+// 	if err := json.Unmarshal(raw, &progress); err != nil {
+// 		return nil, err
+// 	}
+// 	return &coreth.SyncProgress{
+// 		StartingBlock: uint64(progress.StartingBlock),
+// 		CurrentBlock:  uint64(progress.CurrentBlock),
+// 		HighestBlock:  uint64(progress.HighestBlock),
+// 		PulledStates:  uint64(progress.PulledStates),
+// 		KnownStates:   uint64(progress.KnownStates),
+// 	}, nil
+// }
```

Comment:
Generally speaking, we should avoid committing commented out code. Is there a reason we have some here? (and in the rest of this file?)

Reply:
Since this is taken from geth, I find it much easier to know what changes have been made by simply commenting out code that we don't need anymore as opposed to looking through the git history or pulling up the geth codebase to compare and see what's changed.